File: nextjs/src/lib/nflStats.ts
// src/lib/nflStats.ts
//
// The catalog of rankable NflWeeklyStat columns — every numeric stat nflverse
// publishes, which is what the Statistics tab ranks players by.
//
// This is deliberately one list rather than two. The API needs it as a security
// allowlist (the column name is interpolated into raw SQL, so it can never come
// from user input) and the UI needs it for the dropdown. Keeping them as
// separate hand-maintained arrays meant a stat added to one and not the other
// either vanished from the menu or 400'd when picked.
//
// Identity columns, the season/week keys, and the semicolon-joined FG distance
// lists are excluded: none of them are meaningful to rank by.

export interface StatCategory {
  /** NflWeeklyStat column name. Safe to interpolate — it comes from this file. */
  key: string;
  label: string;
  /** Suffix shown after the value, e.g. "yds". Empty for a bare count. */
  unit: string;
  decimals: number;
  group: string;
}

export const STAT_CATEGORIES: StatCategory[] = [
  // ── Fantasy ─────────────────────────────────────────────────────
  { key: 'fantasyPoints',    label: 'Fantasy Points (STD)', unit: 'pts',  decimals: 1, group: 'Fantasy' },
  { key: 'fantasyPointsPpr', label: 'Fantasy Points (PPR)', unit: 'pts',  decimals: 1, group: 'Fantasy' },

  // ── Passing ─────────────────────────────────────────────────────
  { key: 'attempts',               label: 'Pass Attempts',           unit: '',     decimals: 0, group: 'Passing' },
  { key: 'completions',            label: 'Completions',             unit: '',     decimals: 0, group: 'Passing' },
  { key: 'pacr',                   label: 'PACR',                    unit: '',     decimals: 2, group: 'Passing' },
  { key: 'passing10',              label: 'Passing 10',              unit: '',     decimals: 0, group: 'Passing' },
  { key: 'passing16',              label: 'Passing 16',              unit: '',     decimals: 0, group: 'Passing' },
  { key: 'passing20',              label: 'Passing 20',              unit: '',     decimals: 0, group: 'Passing' },
  { key: 'passing2ptConversions',  label: 'Passing 2PT Conversions', unit: '',     decimals: 0, group: 'Passing' },
  { key: 'passing40',              label: 'Passing 40',              unit: '',     decimals: 0, group: 'Passing' },
  { key: 'passingAirYards',        label: 'Air Yards',               unit: 'yds',  decimals: 0, group: 'Passing' },
  { key: 'passingCpoe',            label: 'CPOE',                    unit: '%',    decimals: 1, group: 'Passing' },
  { key: 'passingEpa',             label: 'Passing EPA',             unit: '',     decimals: 1, group: 'Passing' },
  { key: 'passingFirstDowns',      label: 'Pass 1st Downs',          unit: '',     decimals: 0, group: 'Passing' },
  { key: 'passingInterceptions',   label: 'Interceptions',           unit: '',     decimals: 0, group: 'Passing' },
  { key: 'passingTds',             label: 'Passing TDs',             unit: 'TD',   decimals: 0, group: 'Passing' },
  { key: 'passingYards',           label: 'Passing Yards',           unit: 'yds',  decimals: 0, group: 'Passing' },
  { key: 'passingYardsAfterCatch', label: 'YAC',                     unit: 'yds',  decimals: 0, group: 'Passing' },
  { key: 'sackFumbles',            label: 'Sack Fumbles',            unit: '',     decimals: 0, group: 'Passing' },
  { key: 'sackFumblesLost',        label: 'Sack Fumbles Lost',       unit: '',     decimals: 0, group: 'Passing' },
  { key: 'sackYardsLost',          label: 'Sack Yards Lost',         unit: 'yds',  decimals: 0, group: 'Passing' },
  { key: 'sacksSuffered',          label: 'Sacks Taken',             unit: '',     decimals: 0, group: 'Passing' },

  // ── Rushing ─────────────────────────────────────────────────────
  { key: 'carries',               label: 'Carries',                 unit: '',     decimals: 0, group: 'Rushing' },
  { key: 'rushing10',             label: 'Rushing 10',              unit: '',     decimals: 0, group: 'Rushing' },
  { key: 'rushing12',             label: 'Rushing 12',              unit: '',     decimals: 0, group: 'Rushing' },
  { key: 'rushing20',             label: 'Rushing 20',              unit: '',     decimals: 0, group: 'Rushing' },
  { key: 'rushing2ptConversions', label: 'Rushing 2PT Conversions', unit: '',     decimals: 0, group: 'Rushing' },
  { key: 'rushing40',             label: 'Rushing 40',              unit: '',     decimals: 0, group: 'Rushing' },
  { key: 'rushingEpa',            label: 'Rushing EPA',             unit: '',     decimals: 1, group: 'Rushing' },
  { key: 'rushingFirstDowns',     label: 'Rush 1st Downs',          unit: '',     decimals: 0, group: 'Rushing' },
  { key: 'rushingFumbles',        label: 'Rushing Fumbles',         unit: '',     decimals: 0, group: 'Rushing' },
  { key: 'rushingFumblesLost',    label: 'Rushing Fumbles Lost',    unit: '',     decimals: 0, group: 'Rushing' },
  { key: 'rushingTds',            label: 'Rushing TDs',             unit: 'TD',   decimals: 0, group: 'Rushing' },
  { key: 'rushingYards',          label: 'Rushing Yards',           unit: 'yds',  decimals: 0, group: 'Rushing' },

  // ── Receiving ───────────────────────────────────────────────────
  { key: 'airYardsShare',            label: 'Air Yards Share',           unit: '%',    decimals: 1, group: 'Receiving' },
  { key: 'racr',                     label: 'RACR',                      unit: '',     decimals: 2, group: 'Receiving' },
  { key: 'receiving10',              label: 'Receiving 10',              unit: '',     decimals: 0, group: 'Receiving' },
  { key: 'receiving16',              label: 'Receiving 16',              unit: '',     decimals: 0, group: 'Receiving' },
  { key: 'receiving20',              label: 'Receiving 20',              unit: '',     decimals: 0, group: 'Receiving' },
  { key: 'receiving2ptConversions',  label: 'Receiving 2PT Conversions', unit: '',     decimals: 0, group: 'Receiving' },
  { key: 'receiving40',              label: 'Receiving 40',              unit: '',     decimals: 0, group: 'Receiving' },
  { key: 'receivingAirYards',        label: 'Air Yards',                 unit: 'yds',  decimals: 0, group: 'Receiving' },
  { key: 'receivingEpa',             label: 'Rec EPA',                   unit: '',     decimals: 1, group: 'Receiving' },
  { key: 'receivingFirstDowns',      label: 'Rec 1st Downs',             unit: '',     decimals: 0, group: 'Receiving' },
  { key: 'receivingFumbles',         label: 'Receiving Fumbles',         unit: '',     decimals: 0, group: 'Receiving' },
  { key: 'receivingFumblesLost',     label: 'Receiving Fumbles Lost',    unit: '',     decimals: 0, group: 'Receiving' },
  { key: 'receivingTds',             label: 'Receiving TDs',             unit: 'TD',   decimals: 0, group: 'Receiving' },
  { key: 'receivingYards',           label: 'Receiving Yards',           unit: 'yds',  decimals: 0, group: 'Receiving' },
  { key: 'receivingYardsAfterCatch', label: 'YAC',                       unit: 'yds',  decimals: 0, group: 'Receiving' },
  { key: 'receptions',               label: 'Receptions',                unit: '',     decimals: 0, group: 'Receiving' },
  { key: 'targetShare',              label: 'Target Share',              unit: '%',    decimals: 1, group: 'Receiving' },
  { key: 'targets',                  label: 'Targets',                   unit: '',     decimals: 0, group: 'Receiving' },
  { key: 'wopr',                     label: 'WOPR',                      unit: '',     decimals: 2, group: 'Receiving' },

  // ── Defense ─────────────────────────────────────────────────────
  { key: 'def2ptAtts',             label: 'Def 2PT Attempts',           unit: '',     decimals: 0, group: 'Defense' },
  { key: 'def2ptMade',             label: 'Def 2PT Made',               unit: '',     decimals: 0, group: 'Defense' },
  { key: 'defFgBlocks',            label: 'Def FG Blocks',              unit: '',     decimals: 0, group: 'Defense' },
  { key: 'defFumbles',             label: 'Def Fumbles',                unit: '',     decimals: 0, group: 'Defense' },
  { key: 'defFumblesForced',       label: 'Forced Fumbles',             unit: '',     decimals: 0, group: 'Defense' },
  { key: 'defInterceptionYards',   label: 'Def Interception Yards',     unit: 'yds',  decimals: 0, group: 'Defense' },
  { key: 'defInterceptions',       label: 'INTs',                       unit: '',     decimals: 0, group: 'Defense' },
  { key: 'defPassDefended',        label: 'Pass Breakups',              unit: '',     decimals: 0, group: 'Defense' },
  { key: 'defPatBlocks',           label: 'Def PAT Blocks',             unit: '',     decimals: 0, group: 'Defense' },
  { key: 'defPuntBlocks',          label: 'Def Punt Blocks',            unit: '',     decimals: 0, group: 'Defense' },
  { key: 'defQbHits',              label: 'QB Hits',                    unit: '',     decimals: 0, group: 'Defense' },
  { key: 'defSackYards',           label: 'Def Sack Yards',             unit: 'yds',  decimals: 0, group: 'Defense' },
  { key: 'defSacks',               label: 'Sacks',                      unit: '',     decimals: 1, group: 'Defense' },
  { key: 'defSafeties',            label: 'Def Safeties',               unit: '',     decimals: 0, group: 'Defense' },
  { key: 'defTackleAssists',       label: 'Def Tackle Assists',         unit: '',     decimals: 0, group: 'Defense' },
  { key: 'defTacklesForLoss',      label: 'TFL',                        unit: '',     decimals: 1, group: 'Defense' },
  { key: 'defTacklesForLossYards', label: 'Def Tackles For Loss Yards', unit: 'yds',  decimals: 0, group: 'Defense' },
  { key: 'defTacklesSolo',         label: 'Solo Tackles',               unit: '',     decimals: 0, group: 'Defense' },
  { key: 'defTacklesWithAssist',   label: 'Def Tackles With Assist',    unit: '',     decimals: 0, group: 'Defense' },
  { key: 'defTds',                 label: 'Def TDs',                    unit: 'TD',   decimals: 0, group: 'Defense' },

  // ── Kicking ─────────────────────────────────────────────────────
  { key: 'fgAtt',             label: 'FG Attempts',         unit: '',     decimals: 0, group: 'Kicking' },
  { key: 'fgBlocked',         label: 'FG Blocked',          unit: '',     decimals: 0, group: 'Kicking' },
  { key: 'fgBlockedDistance', label: 'FG Blocked Distance', unit: 'yds',  decimals: 0, group: 'Kicking' },
  { key: 'fgLong',            label: 'FG Long',             unit: '',     decimals: 0, group: 'Kicking' },
  { key: 'fgMade',            label: 'FG Made',             unit: '',     decimals: 0, group: 'Kicking' },
  { key: 'fgMade0To19',       label: 'FG Made 0–19',        unit: '',     decimals: 0, group: 'Kicking' },
  { key: 'fgMade20To29',      label: 'FG Made 20–29',       unit: '',     decimals: 0, group: 'Kicking' },
  { key: 'fgMade30To39',      label: 'FG Made 30–39',       unit: '',     decimals: 0, group: 'Kicking' },
  { key: 'fgMade40To49',      label: 'FG Made 40–49',       unit: '',     decimals: 0, group: 'Kicking' },
  { key: 'fgMade50To59',      label: 'FG Made 50–59',       unit: '',     decimals: 0, group: 'Kicking' },
  { key: 'fgMade60Plus',      label: 'FG Made 60 Plus',     unit: '',     decimals: 0, group: 'Kicking' },
  { key: 'fgMadeDistance',    label: 'FG Made Distance',    unit: 'yds',  decimals: 0, group: 'Kicking' },
  { key: 'fgMissed',          label: 'FG Missed',           unit: '',     decimals: 0, group: 'Kicking' },
  { key: 'fgMissed0To19',     label: 'FG Missed 0–19',      unit: '',     decimals: 0, group: 'Kicking' },
  { key: 'fgMissed20To29',    label: 'FG Missed 20–29',     unit: '',     decimals: 0, group: 'Kicking' },
  { key: 'fgMissed30To39',    label: 'FG Missed 30–39',     unit: '',     decimals: 0, group: 'Kicking' },
  { key: 'fgMissed40To49',    label: 'FG Missed 40–49',     unit: '',     decimals: 0, group: 'Kicking' },
  { key: 'fgMissed50To59',    label: 'FG Missed 50–59',     unit: '',     decimals: 0, group: 'Kicking' },
  { key: 'fgMissed60Plus',    label: 'FG Missed 60 Plus',   unit: '',     decimals: 0, group: 'Kicking' },
  { key: 'fgMissedDistance',  label: 'FG Missed Distance',  unit: 'yds',  decimals: 0, group: 'Kicking' },
  { key: 'fgPct',             label: 'FG %',                unit: '%',    decimals: 2, group: 'Kicking' },
  { key: 'gwfgAtt',           label: 'GW FG Attempts',      unit: '',     decimals: 0, group: 'Kicking' },
  { key: 'gwfgBlocked',       label: 'GW FG Blocked',       unit: '',     decimals: 0, group: 'Kicking' },
  { key: 'gwfgDistance',      label: 'GW FG Distance',      unit: 'yds',  decimals: 0, group: 'Kicking' },
  { key: 'gwfgMade',          label: 'GW FG Made',          unit: '',     decimals: 0, group: 'Kicking' },
  { key: 'gwfgMissed',        label: 'GW FG Missed',        unit: '',     decimals: 0, group: 'Kicking' },
  { key: 'patAtt',            label: 'PAT Attempts',        unit: '',     decimals: 0, group: 'Kicking' },
  { key: 'patBlocked',        label: 'PAT Blocked',         unit: '',     decimals: 0, group: 'Kicking' },
  { key: 'patMade',           label: 'PAT Made',            unit: '',     decimals: 0, group: 'Kicking' },
  { key: 'patMissed',         label: 'PAT Missed',          unit: '',     decimals: 0, group: 'Kicking' },
  { key: 'patPct',            label: 'PAT %',               unit: '%',    decimals: 2, group: 'Kicking' },

  // ── Punting ─────────────────────────────────────────────────────
  { key: 'ptAtt',         label: 'Punt Attempts',      unit: '',     decimals: 0, group: 'Punting' },
  { key: 'ptBlocked',     label: 'Punt Blocked',       unit: '',     decimals: 0, group: 'Punting' },
  { key: 'ptDowned',      label: 'Punt Downed',        unit: '',     decimals: 0, group: 'Punting' },
  { key: 'ptFairCaught',  label: 'Punt Fair Caught',   unit: '',     decimals: 0, group: 'Punting' },
  { key: 'ptInside20',    label: 'Punt Inside 20',     unit: '',     decimals: 0, group: 'Punting' },
  { key: 'ptLong',        label: 'Punt Long',          unit: '',     decimals: 0, group: 'Punting' },
  { key: 'ptNetYards',    label: 'Punt Net Yards',     unit: 'yds',  decimals: 0, group: 'Punting' },
  { key: 'ptOutOfBounds', label: 'Punt Out Of Bounds', unit: '',     decimals: 0, group: 'Punting' },
  { key: 'ptReturnTds',   label: 'Punt Return TDs',    unit: 'TD',   decimals: 0, group: 'Punting' },
  { key: 'ptReturnYards', label: 'Punt Return Yards',  unit: 'yds',  decimals: 0, group: 'Punting' },
  { key: 'ptReturned',    label: 'Punt Returned',      unit: '',     decimals: 0, group: 'Punting' },
  { key: 'ptTouchback',   label: 'Punt Touchback',     unit: '',     decimals: 0, group: 'Punting' },
  { key: 'ptYards',       label: 'Punt Yards',         unit: 'yds',  decimals: 0, group: 'Punting' },

  // ── Returns ─────────────────────────────────────────────────────
  { key: 'kickoffReturnYards', label: 'Kickoff Return Yards', unit: 'yds',  decimals: 0, group: 'Returns' },
  { key: 'kickoffReturns',     label: 'Kickoff Returns',      unit: '',     decimals: 0, group: 'Returns' },
  { key: 'puntReturnYards',    label: 'Punt Return Yards',    unit: 'yds',  decimals: 0, group: 'Returns' },
  { key: 'puntReturns',        label: 'Punt Returns',         unit: '',     decimals: 0, group: 'Returns' },
  { key: 'specialTeamsTds',    label: 'Special Teams TDs',    unit: 'TD',   decimals: 0, group: 'Returns' },

  // ── Misc ────────────────────────────────────────────────────────
  { key: 'fumbleRecoveryOpp',      label: 'Fumble Recovery Opponent',       unit: '',     decimals: 0, group: 'Misc' },
  { key: 'fumbleRecoveryOwn',      label: 'Fumble Recovery Own',            unit: '',     decimals: 0, group: 'Misc' },
  { key: 'fumbleRecoveryTds',      label: 'Fumble Recovery TDs',            unit: 'TD',   decimals: 0, group: 'Misc' },
  { key: 'fumbleRecoveryYardsOpp', label: 'Fumble Recovery Yards Opponent', unit: 'yds',  decimals: 0, group: 'Misc' },
  { key: 'fumbleRecoveryYardsOwn', label: 'Fumble Recovery Yards Own',      unit: 'yds',  decimals: 0, group: 'Misc' },
  { key: 'fumblesForcedByOpp',     label: 'Fumbles Forced By Opponent',     unit: '',     decimals: 0, group: 'Misc' },
  { key: 'fumblesLostTotal',       label: 'Fumbles Lost Total',             unit: '',     decimals: 0, group: 'Misc' },
  { key: 'fumblesNotForced',       label: 'Fumbles Not Forced',             unit: '',     decimals: 0, group: 'Misc' },
  { key: 'fumblesOutOfBounds',     label: 'Fumbles Out Of Bounds',          unit: '',     decimals: 0, group: 'Misc' },
  { key: 'fumblesTotal',           label: 'Fumbles Total',                  unit: '',     decimals: 0, group: 'Misc' },
  { key: 'miscYards',              label: 'Misc Yards',                     unit: 'yds',  decimals: 0, group: 'Misc' },
  { key: 'penalties',              label: 'Penalties',                      unit: '',     decimals: 0, group: 'Misc' },
  { key: 'penaltyYards',           label: 'Penalty Yards',                  unit: 'yds',  decimals: 0, group: 'Misc' },
];

/** Distinct groups, in the order the dropdown should show them. */
export const STAT_GROUPS: string[] = [...new Set(STAT_CATEGORIES.map((c) => c.group))];

/** Guards the raw-SQL column interpolation in /api/nfl/leaders. */
export const ALLOWED_STAT_COLS: ReadonlySet<string> = new Set(
  STAT_CATEGORIES.map((c) => c.key),
);

File: python/scripts/common/nflstats.py
"""Loading and upserting nflverse weekly player stats.

Shared by sync_nfl_weekly (one week, in season), load_completed_season (the
finished season each August) and backfill_nfl_seasons (history, run by hand).

COLUMN_MAP is deliberately narrower than what nflverse returns: it lists exactly
the columns NflWeeklyStat keeps. Anything not named here is dropped before it
reaches the database, so adding a column means adding it in both this map and
prisma/schema.prisma.
"""
from __future__ import annotations

import nflreadpy as nfl
import polars as pl

from common import turso

# nflverse source column → NflWeeklyStat column.
COLUMN_MAP: dict[str, str] = {
    # ── Identity ──────────────────────────────────────────────────────
    "game_id": "gameId",
    "player_id": "playerId",
    "player_name": "playerName",
    "player_display_name": "playerDisplayName",
    "position": "position",
    "position_group": "positionGroup",
    "season": "season",
    "week": "week",
    "season_type": "seasonType",
    "team": "team",
    "opponent_team": "opponentTeam",
    # ── Passing ───────────────────────────────────────────────────────
    "completions": "completions",
    "attempts": "attempts",
    "passing_yards": "passingYards",
    "passing_tds": "passingTds",
    "passing_interceptions": "passingInterceptions",
    "sacks_suffered": "sacksSuffered",
    "sack_yards_lost": "sackYardsLost",
    "sack_fumbles": "sackFumbles",
    "sack_fumbles_lost": "sackFumblesLost",
    "passing_air_yards": "passingAirYards",
    "passing_yards_after_catch": "passingYardsAfterCatch",
    "passing_first_downs": "passingFirstDowns",
    "passing_epa": "passingEpa",
    "passing_cpoe": "passingCpoe",
    "passing_2pt_conversions": "passing2ptConversions",
    "pacr": "pacr",
    "passing_10": "passing10",
    "passing_16": "passing16",
    "passing_20": "passing20",
    "passing_40": "passing40",
    # ── Rushing ───────────────────────────────────────────────────────
    "carries": "carries",
    "rushing_yards": "rushingYards",
    "rushing_tds": "rushingTds",
    "rushing_fumbles": "rushingFumbles",
    "rushing_fumbles_lost": "rushingFumblesLost",
    "rushing_first_downs": "rushingFirstDowns",
    "rushing_epa": "rushingEpa",
    "rushing_2pt_conversions": "rushing2ptConversions",
    "rushing_10": "rushing10",
    "rushing_12": "rushing12",
    "rushing_20": "rushing20",
    "rushing_40": "rushing40",
    # ── Receiving ─────────────────────────────────────────────────────
    "receptions": "receptions",
    "targets": "targets",
    "receiving_yards": "receivingYards",
    "receiving_tds": "receivingTds",
    "receiving_fumbles": "receivingFumbles",
    "receiving_fumbles_lost": "receivingFumblesLost",
    "receiving_air_yards": "receivingAirYards",
    "receiving_yards_after_catch": "receivingYardsAfterCatch",
    "receiving_first_downs": "receivingFirstDowns",
    "receiving_epa": "receivingEpa",
    "receiving_2pt_conversions": "receiving2ptConversions",
    "receiving_10": "receiving10",
    "receiving_16": "receiving16",
    "receiving_20": "receiving20",
    "receiving_40": "receiving40",
    "racr": "racr",
    "target_share": "targetShare",
    "air_yards_share": "airYardsShare",
    "wopr": "wopr",
    # ── Defense ───────────────────────────────────────────────────────
    "def_tackles_solo": "defTacklesSolo",
    "def_tackles_with_assist": "defTacklesWithAssist",
    "def_tackle_assists": "defTackleAssists",
    "def_tackles_for_loss": "defTacklesForLoss",
    "def_tackles_for_loss_yards": "defTacklesForLossYards",
    "def_fumbles_forced": "defFumblesForced",
    "def_sacks": "defSacks",
    "def_sack_yards": "defSackYards",
    "def_qb_hits": "defQbHits",
    "def_interceptions": "defInterceptions",
    "def_interception_yards": "defInterceptionYards",
    "def_pass_defended": "defPassDefended",
    "def_tds": "defTds",
    "def_fumbles": "defFumbles",
    "def_safeties": "defSafeties",
    "def_punt_blocks": "defPuntBlocks",
    "def_pat_blocks": "defPatBlocks",
    "def_fg_blocks": "defFgBlocks",
    "def_2pt_atts": "def2ptAtts",
    "def_2pt_made": "def2ptMade",
    # ── Kicking ───────────────────────────────────────────────────────
    "fg_made": "fgMade",
    "fg_att": "fgAtt",
    "fg_missed": "fgMissed",
    "fg_blocked": "fgBlocked",
    "fg_long": "fgLong",
    "fg_pct": "fgPct",
    "fg_made_0_19": "fgMade0To19",
    "fg_made_20_29": "fgMade20To29",
    "fg_made_30_39": "fgMade30To39",
    "fg_made_40_49": "fgMade40To49",
    "fg_made_50_59": "fgMade50To59",
    "fg_made_60_": "fgMade60Plus",
    "fg_missed_0_19": "fgMissed0To19",
    "fg_missed_20_29": "fgMissed20To29",
    "fg_missed_30_39": "fgMissed30To39",
    "fg_missed_40_49": "fgMissed40To49",
    "fg_missed_50_59": "fgMissed50To59",
    "fg_missed_60_": "fgMissed60Plus",
    "fg_made_list": "fgMadeList",
    "fg_missed_list": "fgMissedList",
    "fg_blocked_list": "fgBlockedList",
    "fg_made_distance": "fgMadeDistance",
    "fg_missed_distance": "fgMissedDistance",
    "fg_blocked_distance": "fgBlockedDistance",
    "pat_made": "patMade",
    "pat_att": "patAtt",
    "pat_missed": "patMissed",
    "pat_blocked": "patBlocked",
    "pat_pct": "patPct",
    "gwfg_made": "gwfgMade",
    "gwfg_att": "gwfgAtt",
    "gwfg_missed": "gwfgMissed",
    "gwfg_blocked": "gwfgBlocked",
    "gwfg_distance": "gwfgDistance",
    # ── Punting ───────────────────────────────────────────────────────
    "pt_att": "ptAtt",
    "pt_blocked": "ptBlocked",
    "pt_long": "ptLong",
    "pt_yards": "ptYards",
    "pt_inside_20": "ptInside20",
    "pt_out_of_bounds": "ptOutOfBounds",
    "pt_downed": "ptDowned",
    "pt_touchback": "ptTouchback",
    "pt_fair_caught": "ptFairCaught",
    "pt_returned": "ptReturned",
    "pt_return_yards": "ptReturnYards",
    "pt_return_tds": "ptReturnTds",
    "pt_net_yards": "ptNetYards",
    # ── Returns / misc ────────────────────────────────────────────────
    "special_teams_tds": "specialTeamsTds",
    "misc_yards": "miscYards",
    "fumble_recovery_own": "fumbleRecoveryOwn",
    "fumble_recovery_yards_own": "fumbleRecoveryYardsOwn",
    "fumble_recovery_opp": "fumbleRecoveryOpp",
    "fumble_recovery_yards_opp": "fumbleRecoveryYardsOpp",
    "fumble_recovery_tds": "fumbleRecoveryTds",
    "penalties": "penalties",
    "penalty_yards": "penaltyYards",
    "fumbles_forced_by_opp": "fumblesForcedByOpp",
    "fumbles_not_forced": "fumblesNotForced",
    "fumbles_out_of_bounds": "fumblesOutOfBounds",
    "fumbles_total": "fumblesTotal",
    "fumbles_lost_total": "fumblesLostTotal",
    "punt_returns": "puntReturns",
    "punt_return_yards": "puntReturnYards",
    "kickoff_returns": "kickoffReturns",
    "kickoff_return_yards": "kickoffReturnYards",
    # ── Fantasy ───────────────────────────────────────────────────────
    "fantasy_points": "fantasyPoints",
    "fantasy_points_ppr": "fantasyPointsPpr",
    # ── Other ────────────────────────────────────────────────────────────
    "headshot_url": "headshot",
}

# Columns stored as TEXT — everything else is cast to Float64.
TEXT_COLUMNS: frozenset[str] = frozenset({
    "fgBlockedList", "fgMadeList", "fgMissedList", "gameId", "headshot",
    "opponentTeam", "playerDisplayName", "playerId", "playerName", "position",
    "positionGroup", "seasonType", "team",
})

# Part of the unique key, so never included in the ON CONFLICT update list.
KEY_COLUMNS: frozenset[str] = frozenset({"season", "week", "playerId"})


def _narrow(df: pl.DataFrame) -> pl.DataFrame:
    """Selects and renames the stored columns, then normalises numeric types."""
    available = [c for c in COLUMN_MAP if c in df.columns]
    df = df.select(available).rename({c: COLUMN_MAP[c] for c in available})
    df = df.filter(pl.col("playerId").is_not_null())

    return df.with_columns([
        pl.col(c).cast(pl.Float64, strict=False)
        for c in df.columns
        if c not in TEXT_COLUMNS and df[c].dtype != pl.Utf8
    ])


def load_seasons(seasons: list[int]) -> pl.DataFrame:
    """Loads every published week for the given seasons."""
    return _narrow(nfl.load_player_stats(seasons))


def load_latest_week(season: int) -> tuple[pl.DataFrame, int]:
    """Loads only the most recent week nflverse has published for `season`.

    Returns the rows and the week number. One fetch serves both — nflverse ships
    the whole season in a single file, so re-reading it to find the max week
    would double the download for no gain.
    """
    df = nfl.load_player_stats([season])
    week = df.select(pl.col("week").max()).item()
    return _narrow(df.filter(pl.col("week") == week)), week


def rows_per_season() -> dict[int, int]:
    """How many rows NflWeeklyStat already holds for each season.

    Used to skip a season that is already loaded. Re-uploading one costs
    minutes and achieves nothing when the row count already matches what
    nflverse is offering.
    """
    rows = turso.query(
        'SELECT season, COUNT(*) AS n FROM "NflWeeklyStat" GROUP BY season'
    )
    return {int(r["season"]): int(r["n"]) for r in rows}


def upsert(df: pl.DataFrame) -> int:
    """Upserts every row keyed on (season, week, playerId). Returns rows written."""
    cols = df.columns
    col_names = ", ".join(f'"{c}"' for c in cols)
    placeholders = ", ".join("?" for _ in cols)
    updates = ", ".join(
        f'"{c}" = excluded."{c}"' for c in cols if c not in KEY_COLUMNS
    )

    sql = (
        f"INSERT INTO NflWeeklyStat (id, {col_names}) "
        f"VALUES (lower(hex(randomblob(16))), {placeholders}) "
        f'ON CONFLICT (season, week, "playerId") DO UPDATE SET {updates}'
    )

    rows = [[row[c] for c in cols] for row in df.to_dicts()]
    return turso.execute_chunked(sql, rows, label="stat rows")


def truncate() -> None:
    """Empties NflWeeklyStat.

    No scheduled job calls this any more. The annual August load used to
    truncate before reloading a rolling window of seasons; it now appends the
    finished season and deletes nothing, because the card pool is built from
    every season in the table and a wipe threw away history the game depends on
    — including a 1999+ backfill that takes four hours to rebuild.

    Kept as a maintenance tool for a deliberate, manual rebuild. If you are
    about to call it, you are about to lose every season not in whatever you
    load next.
    """
    print("Truncating NflWeeklyStat...")
    turso.execute_one('DELETE FROM "NflWeeklyStat"')
    print("  Table cleared.")
